[iPadOS 18] Unable to include header row when copy and pasting table from confluence to Notes

whsieh opened the pull request:
#### 1b4675b62ba332657e2461a9e45e6068c391a3df
<pre>
[iPadOS 18] Unable to include header row when copy and pasting table from confluence to Notes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281230">https://bugs.webkit.org/show_bug.cgi?id=281230</a>
<a href="https://rdar.apple.com/137088450">rdar://137088450</a>

Reviewed by Aditya Keerthi.

On Confluence, table header content is styled with `-webkit-user-select: none;`, which (in addition
to text selection) normally prevents the text from being serialized when writing content to the
system pasteboard when copying. This policy decision is overridden by a site-specific quirk (added
in 265939@main) which makes us *not* ignore `user-select: none;` when copying on this website. This
quirk works by detecting Confluence via the presence of a `meta` tag with `confluence-request-time`
and making `HTMLConverter` and `serializePreservingVisualAppearance` not ignore content inside of
`-webkit-user-select: none;` subtrees. When pasting into Notes, Notes asks for attributed string
data, which contains this `user-select: none;` content when copied from Confluence.

On iPad running iOS 18 and later, however, we no longer write `NSAttributedString` data directly to
the pasteboard when copying in WebKit apps, and instead defer to `UIFoundation` to perform this
conversion only if needed when pasting. Underneath the hood, this ultimately still uses WebKit
through `nsattributedstringagent`, which receives either web archive or HTML markup data and
generates an equivalent `NSAttributedString` / RTF data. Because the offscreen web view in
`nsattributedstringagent` only converts web archive data to an attributed string (and importantly,
doesn&apos;t load Confluence), it doesn&apos;t enable the aforementioned quirk, and so table header content
ends up getting dropped from the converted attributed string.

To fix this, we adjust `-[WKWebView _getContentsAsAttributedStringWithCompletionHandler:]` (and all
upstream API/SPIs) to *not* ignore `-webkit-user-select: none;` when converting webpage content
into attributed strings.

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::selectionAsAttributedString):
* Source/WebCore/editing/cocoa/HTMLConverter.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(WebCore::attributedString):

Add more plumbing for the `IgnoreUserSelectNone` flag, through attributed string serialization
codepaths.

* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::dataSelectionForPasteboard):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:

Implement the fix by passing `IgnoreUserSelectNone::No` into `attributedString()`. See above for
more details.

(WebKit::WebPage::getContentsAsAttributedString):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::handleSelectionServiceClick):
* Source/WebKitLegacy/mac/WebCoreSupport/WebSelectionServiceController.mm:
(WebSelectionServiceController::handleSelectionServiceClick):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _legacyAttributedStringFrom:offset:to:offset:]):
(-[WebHTMLView attributedString]):
(-[WebHTMLView selectedAttributedString]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST(WKWebView, AttributedStringIncludesUserSelectNoneContent)):

Add an API test to exercise the new behavior (see above for more details).

Canonical link: <a href="https://commits.webkit.org/284993@main">https://commits.webkit.org/284993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84befa1cb136168c32a1fa55d0cb0fe8fe4ad55c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50501 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75194 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73204 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56216 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14690 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74154 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45900 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61286 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42566 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18738 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20634 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64470 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76916 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18283 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15365 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63906 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15750 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12035 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5676 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46302 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1081 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47373 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47115 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
```diff
@@ -562,3 +562,12 @@ - (NSAttributedString *)_contentsAsAttributedString
+TEST(WKWebView, AttributedStringIncludesUserSelectNoneContent)
+{
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300)]);
+    [webView synchronouslyLoadHTMLString:@"<body><p style='-webkit-user-select: none;'>Hello</p></body>"];
```
Hm... doesn't this mean -webkit-user-select: none content will once again always appear in the pasted content. I feel like a better fix would be to detect whether the content came from confluence by checking class names, id, etc... in the pasted content.

rniwa replied:
We talked about this in person. The content web archive / HTML copied into pasteboard won't contain `-webkit-user-select: none` content on websites other than Confluence due to this quirk being specific to Confluence so we wouldn't end up re-introducing `-webkit-user-select: none` content elsewhere.

whsieh committed:
[iPadOS 18] Unable to include header row when copy and pasting table from confluence to Notes
https://bugs.webkit.org/show_bug.cgi?id=281230
rdar://137088450

Reviewed by Aditya Keerthi.

On Confluence, table header content is styled with `-webkit-user-select: none;`, which (in addition
to text selection) normally prevents the text from being serialized when writing content to the
system pasteboard when copying. This policy decision is overridden by a site-specific quirk (added
in 265939@main) which makes us *not* ignore `user-select: none;` when copying on this website. This
quirk works by detecting Confluence via the presence of a `meta` tag with `confluence-request-time`
and making `HTMLConverter` and `serializePreservingVisualAppearance` not ignore content inside of
`-webkit-user-select: none;` subtrees. When pasting into Notes, Notes asks for attributed string
data, which contains this `user-select: none;` content when copied from Confluence.

On iPad running iOS 18 and later, however, we no longer write `NSAttributedString` data directly to
the pasteboard when copying in WebKit apps, and instead defer to `UIFoundation` to perform this
conversion only if needed when pasting. Underneath the hood, this ultimately still uses WebKit
through `nsattributedstringagent`, which receives either web archive or HTML markup data and
generates an equivalent `NSAttributedString` / RTF data. Because the offscreen web view in
`nsattributedstringagent` only converts web archive data to an attributed string (and importantly,
doesn't load Confluence), it doesn't enable the aforementioned quirk, and so table header content
ends up getting dropped from the converted attributed string.

To fix this, we adjust `-[WKWebView _getContentsAsAttributedStringWithCompletionHandler:]` (and all
upstream API/SPIs) to *not* ignore `-webkit-user-select: none;` when converting webpage content
into attributed strings.

* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::selectionAsAttributedString):
* Source/WebCore/editing/cocoa/HTMLConverter.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(WebCore::attributedString):

Add more plumbing for the `IgnoreUserSelectNone` flag, through attributed string serialization
codepaths.

* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::dataSelectionForPasteboard):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:

Implement the fix by passing `IgnoreUserSelectNone::No` into `attributedString()`. See above for
more details.

(WebKit::WebPage::getContentsAsAttributedString):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::handleSelectionServiceClick):
* Source/WebKitLegacy/mac/WebCoreSupport/WebSelectionServiceController.mm:
(WebSelectionServiceController::handleSelectionServiceClick):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _legacyAttributedStringFrom:offset:to:offset:]):
(-[WebHTMLView attributedString]):
(-[WebHTMLView selectedAttributedString]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST(WKWebView, AttributedStringIncludesUserSelectNoneContent)):

Add an API test to exercise the new behavior (see above for more details).

Canonical link: https://commits.webkit.org/284993@main
